### PR TITLE
Resync should not send operations without sequence number

### PR DESCRIPTION
**Pull request by dnhatn**
Primary-replica resync in a mixed-cluster between 6.x and 5.6 can send operations without sequence number to a replica which already processed operations with sequence number. This leads to the failure of that replica for we trip [the sequence number assertion](https://github.com/elastic/elasticsearch/blob/58e885ca8a6966ae1cf402a53333ca69caf9fdc7/server/src/main/java/org/elasticsearch/index/seqno/SequenceNumbers.java#L116) when writing resync operations without sequence number to translog.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/index/translog/Translog.java [hunk 1]
```diff
@@ -683,14 +683,8 @@ private Snapshot newMultiSnapshot(TranslogSnapshot[] snapshots) throws IOExcepti
...
-        assert readLock.isHeldByCurrentThread() || writeLock.isHeldByCurrentThread() :
-            "callers of readersAboveMinSeqNo must hold a lock: readLock ["
-                + readLock.isHeldByCurrentThread() + "], writeLock [" + readLock.isHeldByCurrentThread() + "]";
-        return Stream.concat(readers.stream(), Stream.of(current))
-            .filter(reader -> {
-                final long maxSeqNo = reader.getCheckpoint().maxSeqNo;
-                return maxSeqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || maxSeqNo >= minSeqNo;
-            });
+        assert readLock.isHeldByCurrentThread();
+        return Stream.concat(readers.stream(), Stream.of(current)).filter(reader -> reader.getCheckpoint().maxSeqNo >= minSeqNo);
```
Comment (dnhatn) [inline server/src/main/java/org/elasticsearch/index/translog/Translog.java]: This method is consumed by peer-recovery and resync. Peer-recovery already excluded operations without sequence numbers (in #27580); therefore we can safely exclude readers without sequence numbers.

**Comment by bleskes** [comment]
@dnhatn I think I'm missing something. The resync is in charge of verifying operations above the global checkpoint. We currently starting to sync from [the global checkpoint + 1](https://github.com/elastic/elasticsearch/blob/master/server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java#L87) , which seems to only be negative if the global checkpoint is uknown on the new primary. I think we talked before (although I don't remember the details) about bootstrapping the global checkpoint in these cases. I think the right solution is there (and also asserting we always have a starting sequence number for a resync). Is there any reason for you to go down the route of explicit filtering rather than checking the starting sequence number?

**Comment by dnhatn** [comment]
@bleske Thank you for looking :)

Bootstrapping the global checkpoint would work too - I can make that change. I did not have any preference for my implementation, merely followed the pattern we did in #27580. Moreover, if we have an assigned global checkpoint then we will never use operations without seqno; therefore I think we should exclude them from translog snapshot.

**Comment by bleskes** [comment]
> Bootstrapping the global checkpoint would work too - I can make that change. I did not have any preference for my implementation, merely followed the pattern we did in #27580

I'm not sure I follow the reference to the PR as it's about flushing, but that aside - anything we can do to make the situation as "normal" as possible the better it is. Under normal operation (I.e., no BWC mode) we always have a global checkpoint, so that's why I favor that direction.

**Comment by ywelsch** [comment]
There is another related issue here: https://discuss.elastic.co/t/extremely-large-translog-files-per-shard-in-elasticsearch-6-2-4/174041/5 as a gcp of UNASSIGNED_SEQ_NO can cause a resync to send the full translog, which can cause high loads during a rolling upgrade.

> Bootstrapping the global checkpoint would work too - I can make that change.

I'm a little uncomfortable making that change in 6.7.1, as it might have a broader impact, especially in the presence of 5.x indices with no sequence numbers. I wonder if we can have the more contained change here for now to be backported to 6.7 and follow-up with a more comprehensive investigation of getting rid of UNASSIGNED_SEQ_NO in various places as we know that in 7.x all docs will have sequence numbers. WDYT @bleskes?

**Comment by bleskes** [comment]
> WDYT @bleskes?
+1

**Comment by dnhatn** [comment]
@bleskes and @ywelsch This is ready again. Can you please have another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 2]
```diff
@@ -246,11 +246,12 @@ protected void doRun() throws Exception {
             Translog.Operation operation;
             while ((operation = snapshot.next()) != null) {
                 final long seqNo = operation.seqNo();
-                if (startingSeqNo >= 0 &&
-                    (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo)) {
+                // do not send operations if the global checkpoint is still unassigned.
+                if (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || startingSeqNo < 0 || seqNo < startingSeqNo) {
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: I wonder if this is equivalent to the weaker `if (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo) {`, i.e., we could just remove `startingSeqNo >= 0` from the condition.

If the global checkpoint is `SequenceNumbers.UNASSIGNED_SEQ_NO`, I would assume that most operations (except a few that might have happened before global checkpoint has advanced) would be `SequenceNumbers.UNASSIGNED_SEQ_NO`.

**Comment by dnhatn** [comment]
@ywelsch I pushed https://github.com/elastic/elasticsearch/pull/40433/commits/61ad34d3e4a8ffa9d09153b4845f91f5ebb61c6a to use a weaker form which won't send operations without sequence numbers. Can you have another look? Thank you!

File: server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java [hunk 3]
```diff
@@ -246,11 +246,12 @@ protected void doRun() throws Exception {
             Translog.Operation operation;
             while ((operation = snapshot.next()) != null) {
                 final long seqNo = operation.seqNo();
-                if (startingSeqNo >= 0 &&
-                    (seqNo == SequenceNumbers.UNASSIGNED_SEQ_NO || seqNo < startingSeqNo)) {
+                // do not send operations if the global checkpoint is still unassigned.
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/index/shard/PrimaryReplicaSyncer.java]: this comment needs to be updated or removed

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
Thanks @bleskes and @ywelsch.

**Pull request merged**